azurecompute-arm/src/main/java/org/jclouds/azurecompute/arm/compute/AzureComputeServiceAdapter.java

```diff
@@ -0,0 +1,345 @@
+           final String group, final String name, final Template template) {
+
+      final String location = template.getLocation().getId();
+      DeploymentTemplateBuilder deploymentTemplateBuilder = new DeploymentTemplateBuilder(group, name, template);
+
+      final String loginUser = deploymentTemplateBuilder.getLoginUserUsername();
+      final String loginPassword = deploymentTemplateBuilder.getLoginPassword();
+
+      DeploymentBody deploymentTemplateBody =  deploymentTemplateBuilder.getDeploymentTemplate();
+
+      DeploymentProperties properties = DeploymentProperties.create(deploymentTemplateBody);
+      Gson gson = new GsonBuilder().disableHtmlEscaping().create();
+      org.jclouds.json.Json json = new GsonWrapper(gson);
+
+      final String deploymentTemplate = UrlEscapers.urlFormParameterEscaper().escape(json.toJson(properties));
+
+      logger.debug("Deployment created with name: %s", name);
+
+      final Set<VMDeployment> deployments = Sets.newHashSet();
+      if (!retry(new Predicate<String>() {
+         @Override
+         public boolean apply(final String name) {
+            runningNumber++;
+
+
+            ResourceGroupApi resourceGroupApi = api.getResourceGroupApi();
+            ResourceGroup resourceGroup = resourceGroupApi.get(getGroupId());
+            String resourceGroupName;
+
+            if (resourceGroup == null){
+               final Map<String, String> tags = ImmutableMap.of("description", "jClouds managed VMs");
+               resourceGroupName = resourceGroupApi.create(getGroupId(), location, tags).name();
+            } else {
+               resourceGroupName = resourceGroup.name();
+            }
+
+            Deployment deployment = api.getDeploymentApi(resourceGroupName).create(name, deploymentTemplate);
+
+            if (deployment != null) {
+               VMDeployment vmDeployment = new VMDeployment();
+               vmDeployment.deployment = deployment;
+               deployments.add(vmDeployment);
+            } else {
+               logger.debug("Failed to create deployment!");
+            }
+            return !deployments.isEmpty();
+         }
+      }, azureComputeConstants.operationTimeout(), 1, SECONDS).apply(name)) {
+         final String illegalStateExceptionMessage = format("Deployment %s was not created within %sms so it will be destroyed.",
+                 name, azureComputeConstants.operationTimeout());
+         logger.warn(illegalStateExceptionMessage);
+
+         throw new IllegalStateException(illegalStateExceptionMessage);
+      }
+
+      final VMDeployment deployment = deployments.iterator().next();
+
+
+      return new NodeAndInitialCredentials<VMDeployment>(deployment, name,
+              LoginCredentials.builder().user(loginUser).identity(loginUser).password(loginPassword).authenticateSudo(true).build());
+   }
+
+   @Override
+   public Iterable<VMSize> listHardwareProfiles() {
+      return api.getVMSizeApi(getLocation()).list();
```

Comment:
This must return the hardware profiles for *all* locations, and how it works must be coded in conjunction with the function that transforms the returned elements into a jclouds `Hardware`

* If a hardawre profile does not exist in all locations, then when transforming it, the `location` field must be set.
* if a hardware profile is global and exists in all locations, then when transforming it, the `location` field must be left null.

Take this into account so you return here something that gives the transformation function enough context to implement this contract. You can create just an intermediate object here and use it to represent hardware profiles. This object does not have to be an Azure API object, but any object that is convenient; it will never be exposed to users.